File: typescript/src/application/usecase/order/newContributionOrderUsecase.ts
import Decimal from "decimal.js";
import { AccountRepository } from "../../repository/accountRepository.js";
import { MarketPriceRepository } from "../../repository/marketPriceRepository.js";
import { PortfolioRepository } from "../../repository/portfolioRepository.js";
import { PortfolioService } from "../../service/portfolioService.js";
import { AppConstants } from "../../../domain/appConstants.js";
import { UserId } from "../../../domain/userId.js";

export interface NewContributionOrderUsecaseInput {
  userId: UserId;
  amount: Decimal;
}

export class NewContributionOrderUsecaseException extends Error {}
export class NewContributionUserAlreadyExistsException extends NewContributionOrderUsecaseException {
  constructor() {
    super("user already exists");
  }
}
export class NewContributionAmountTooSmallException extends NewContributionOrderUsecaseException {
  constructor() {
    super("amount too small");
  }
}

export class NewContributionOrderUsecase {
  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly portfolioRepository: PortfolioRepository,
    private readonly marketPriceRepository: MarketPriceRepository,
  ) {}

  async run(input: NewContributionOrderUsecaseInput): Promise<void> {
    if (input.amount.lessThan(AppConstants.minOperationAmount)) {
      throw new NewContributionAmountTooSmallException();
    }
    const exists = await this.accountRepository.exists(input.userId);
    if (exists) {
      throw new NewContributionUserAlreadyExistsException();
    }
    const portfolio = await this.portfolioRepository.get();
    const prices = await this.marketPriceRepository.all();
    const account = PortfolioService.allocateNew(input.amount, portfolio, prices);
    await this.accountRepository.upsert(input.userId, account);
  }
}

File: golang/internal/application/usecase/order/new_contribution_order_usecase.go
package order

import (
	"errors"

	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/application/service"
	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

var (
	ErrNewContributionUserAlreadyExists = errors.New("user already has account")
	ErrNewContributionAmountTooSmall    = errors.New("amount is too small")
)

type NewContributionOrderUsecaseInput struct {
	UserId domain.UserId
	Amount decimal.Decimal
}

type NewContributionOrderUsecase struct {
	accountRepo       repository.AccountRepository
	portfolioRepo   repository.PortfolioRepository
	marketPriceRepo repository.MarketPriceRepository
}

func NewNewContributionOrderUsecase(
	accountRepo repository.AccountRepository,
	portfolioRepo repository.PortfolioRepository,
	marketPriceRepo repository.MarketPriceRepository,
) *NewContributionOrderUsecase {
	return &NewContributionOrderUsecase{
		accountRepo:       accountRepo,
		portfolioRepo:   portfolioRepo,
		marketPriceRepo: marketPriceRepo,
	}
}

func (u *NewContributionOrderUsecase) Run(input NewContributionOrderUsecaseInput) error {
	if input.Amount.LessThan(domain.MinOperationAmount) {
		return ErrNewContributionAmountTooSmall
	}

	exists, err := u.accountRepo.Exists(input.UserId)
	if err != nil {
		return err
	}
	if exists {
		return ErrNewContributionUserAlreadyExists
	}

	portfolio, err := u.portfolioRepo.Get()
	if err != nil {
		return err
	}

	prices, err := u.marketPriceRepo.All()
	if err != nil {
		return err
	}

	account, err := service.AllocateNew(input.Amount, portfolio, prices)
	if err != nil {
		return err
	}

	return u.accountRepo.Upsert(input.UserId, account)
}
